packages/env/lib/config/parse-config.js

```diff
@@ -3,206 +3,421 @@
+		getEnvironmentVarOverrides( cacheDirectoryPath );
+
+	// Merge all of our configs so that we have a complete object
+	// containing the desired options in order of precedence.
+	return mergeConfigs(
+		defaultConfig,
+		localConfig ?? {},
+		overrideConfig ?? {},
+		environmentVarOverrides
+	);
+}
+
+/**
+ * Gets the path to the config file.
+ *
+ * @param {string} configDirectoryPath The path to the directory containing config files.
+ * @param {string} type                The type of config file we're interested in: 'local' or 'override'.
+ *
+ * @return {string} The path to the config file.
  */
-const HOME_PATH_PREFIX = `~${ path.sep }`;
+function getConfigFilePath( configDirectoryPath, type = 'local' ) {
+	let fileName;
+	switch ( type ) {
+		case 'local': {
+			fileName = '.wp-env.json';
+			break;
+		}
+
+		case 'override': {
+			fileName = '.wp-env.override.json';
+			break;
+		}
+
+		default: {
+			throw new Error( `Invalid config file type "${ type }.` );
+		}
+	}
+
+	return path.resolve( configDirectoryPath, fileName );
+}
 
 /**
- * Parses a config object. Takes environment-level configuration in the format
- * specified in .wp-env.json, validates it, and converts it into the format used
- * internally. For example, `plugins: string[]` will be parsed into
- * `pluginSources: WPSource[]`.
+ * Gets the default config that can be overridden.
  *
- * @param {Object} config                    A config object to validate.
+ * @param {string} configDirectoryPath        A path to the config file's directory.
  * @param {Object} options
- * @param {string} options.workDirectoryPath Path to the work directory located in ~/.wp-env.
- * @return {WPServiceConfig} Parsed environment-level configuration.
+ * @param {string} options.shouldInferType    Indicates whether or not we should infer the type of project wp-env is being used in.
+ * @param {string} options.cacheDirectoryPath Path to the work directory located in ~/.wp-env.
+ *
+ * @return {Promise<WPServiceConfig>} The default config object.
  */
-module.exports = async function parseConfig( config, options ) {
-	return {
-		port: config.port,
-		phpVersion: config.phpVersion,
-		coreSource: includeTestsPath(
-			await parseCoreSource( config.core, options ),
-			options
-		),
-		pluginSources: config.plugins.map( ( sourceString ) =>
-			parseSourceString( sourceString, options )
-		),
-		themeSources: config.themes.map( ( sourceString ) =>
-			parseSourceString( sourceString, options )
-		),
-		config: config.config,
-		mappings: Object.entries( config.mappings ).reduce(
-			( result, [ wpDir, localDir ] ) => {
-				const source = parseSourceString( localDir, options );
-				result[ wpDir ] = source;
-				return result;
+async function getDefaultConfig(
+	configDirectoryPath,
+	{ shouldInferType, cacheDirectoryPath }
+) {
+	// Our default config should try to infer what type of project
+	// this is in order to automatically map the current directory.
+	const type = shouldInferType
```

Comment:
```suggestion
	const detectedType = shouldInferType
```

just to make it more clear that `type` might not apply in every situation

Reply:
Agreed